Remora.Rest/Json/Internal/BoundDataObjectConverter.cs

```diff
@@ -0,0 +1,272 @@
+            );
+
+            if (property.CanWrite)
+            {
+                // If a property is writable, it can be *read* from JSON.
+                readProperties.Add(data);
+            }
+
+            if (property.CanWrite || source.IncludesReadOnlyProperty(property))
+            {
+                // Any property that is writable and not excluded due to being read-only,
+                // can be *written* to JSON.
+                writeProperties.Add(data);
+            }
+        }
+
+        _writeProperties = writeProperties.ToArray();
+        _readProperties = readProperties.ToArray();
+
+        _readPropertiesByName = readProperties
+            .SelectMany(p => p.ReadNames.Select((n, i) => (IsPrimary: i == 0, n, DTOProperty: p)))
+            .ToDictionary(x => x.n, x => (x.IsPrimary, x.DTOProperty));
+    }
+
+    private static JsonSerializerOptions GetPropertyConverterOptions(JsonSerializerOptions options, JsonConverter? converter)
+    {
+        if (converter == null)
+        {
+            return options;
+        }
+
+        var cloned = new JsonSerializerOptions(options);
+        cloned.Converters.Insert(0, converter);
+
+        return cloned;
+    }
+
+    /// <inheritdoc />
+    public override TInterface Read
+    (
+        ref Utf8JsonReader reader,
+        Type typeToConvert,
+        JsonSerializerOptions options
+    )
+    {
+        if (reader.TokenType != JsonTokenType.StartObject ||
+            !reader.Read())
+        {
+            DataObjectConverterShared.ThrowJsonException();
+        }
+
+        var readProperties = _readProperties;
+
+        // CMBK: Considering to cache this array and reuse it.
+        // F.e. in a thread-static field. This method isn't async anyways.
+
+        // Avoid creating a dictionary and fill the arguments array directly
+        var constructorArguments = new object?[readProperties.Length];
+
+        // We use "Missing" as a sentinel to indicate that a value hasn't
+        // been seen yet without needing to allocate additional memory.
+        constructorArguments.AsSpan().Fill(DataObjectConverterShared.Missing);
+
+        while (reader.TokenType != JsonTokenType.EndObject)
+        {
+            var propertyName = reader.GetString();
+            if (!reader.Read())
+            {
+                DataObjectConverterShared.ThrowJsonException();
+            }
+
+            Debug.Assert(propertyName != null, "Expected property names to never be null.");
+            var (isPrimaryChoice, dtoProperty) = GetReadPropertyInfo(propertyName);
+
+            if (dtoProperty is null)
+            {
+                if (!_allowExtraProperties)
+                {
+                    DataObjectConverterShared.ThrowJsonException();
+                }
+
+                // No matching property - we'll skip it
+                if (!reader.TrySkip())
+                {
+                    DataObjectConverterShared.ThrowJsonException("Couldn't skip elements.");
```

Comment:
Is there a particular reason for adding a throw helper? It's not used anywhere else in the library, and I'd prefer not to mix approaches if we can avoid it.

Reply:
Not particularly beyond preference, I'll revert these.